torch/csrc/utils/python_dispatch.cpp

```diff
@@ -0,0 +1,162 @@
+#include <torch/csrc/utils/python_dispatch.h>
+#include <torch/csrc/jit/frontend/function_schema_parser.h>
+
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/ATen.h>
+
+#include <pybind11/operators.h>
+
+#include <iostream>
+
+namespace py = pybind11;
+
+namespace torch {
+namespace impl {
+namespace dispatch {
+
+c10::optional<c10::DispatchKey> parseDispatchKey(const std::string& k) {
+  static std::unordered_map<std::string, c10::DispatchKey> key_map = {
+    {"cpu", c10::DispatchKey::CPUTensorId},
+    {"cuda", c10::DispatchKey::CUDATensorId},
+    {"xla", c10::DispatchKey::XLATensorId},
+    {"autograd", c10::DispatchKey::VariableTensorId},
+    {"", c10::DispatchKey::Undefined},
+  };
+  auto it = key_map.find(k);
+  TORCH_CHECK(it != key_map.end(), "could not parse ", k);
+  if (it->second == c10::DispatchKey::Undefined) {
```

Comment:
It's because I'm a lazy bastard. If I directly store nullopt in the map, then I have to `make_optional` every other entry in the map (to get the types to line up).

Undefined already exists as the "nullopt" element for DispatchKey. For example, this existing comment:

```
  // ~~~~~~~~~~~~~~~~~~~~~~~~~~ UNDEFINED ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ //
  // This is not a "real" tensor id, but it exists to give us a "nullopt"
  // element we can return for cases when a DispatchKeySet contains no elements.
  // You can think a more semantically accurate definition of DispatchKey is:
  //
  //    using DispatchKey = optional<RealDispatchKey>
  //
  // and Undefined == nullopt.  We didn't actually represent
  // it this way because optional<RealDispatchKey> would take two
  // words, when DispatchKey fits in eight bits.
```

So, technically, when you say `optional<DispatchKey>`, this is actually `optional<optional<RealDispatchKey>>`, which is rarely what you want. But we have a similar situation with Tensor, which also has a nullopt element (the default constructor), but we still opt to say `optional<Tensor>` in some cases to be more explicit about the null-ness (and the default constructor is illegal here). That's kind of what's going on here.